rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -171,51 +203,218 @@ impl MultisigCheckpointSyncer {
                         root = format!("{:#x}", root),
                         signature_count = signature_count,
+                        cumulative_weight = cumulative_weight,
                         "Found signed checkpoint"
                     );
 
-                    // If we've hit a quorum, create a MultisigSignedCheckpoint
-                    if signature_count >= threshold {
+                    // If we've hit a quorum in weight, create a MultisigSignedCheckpoint
+                    if cumulative_weight >= threshold_weight {
+                        // to conform to the onchain ordering of the set by address
+                        signed_checkpoints.sort_by_key(|sc| {
+                            self.fetch_or_recover(sc, &mut recovered_addresses)
+                                .ok()
+                                .and_then(|signer| {
+                                    weighted_validators
+                                        .iter()
+                                        .position(|vw| vw.validator == signer)
+                                })
+                                .map(|pos| (false, pos))
+                                .unwrap_or((true, 0)) // this shouldn't cause a panic because returning (true, 0) means that the validator is not in the weighted_validators list
+                        });
                         let checkpoint: MultisigSignedCheckpoint = signed_checkpoints.try_into()?;
                         debug!(checkpoint=?checkpoint, "Fetched multisig checkpoint");
                         return Ok(Some(checkpoint));
                     }
                 } else {
                     debug!(
-                        validator = format!("{:#x}", validator),
+                        validator = format!("{:#x}", vw.validator),
                         index = index,
                         "Unable to find signed checkpoint"
                     );
                 }
             } else {
-                debug!(%validator, "Unable to find checkpoint syncer");
+                debug!(%vw.validator, "Unable to find checkpoint syncer");
                 continue;
             }
         }
         debug!("No quorum checkpoint found for message");
         Ok(None)
     }
+
+    // this function fetches the highest index that meets the threshold weight
+    // it assumes sorted_indices is sorted in descending order of index
+    // it doesn't assume that the indices contain all validators in weighted_validators
+    fn fetch_highest_quorum_index(
+        &self,
+        weighted_validators: &[ValidatorWithWeight],
+        threshold_weight: Weight,
+        sorted_indices: &[(H256, u32)],
+    ) -> Option<u32> {
+        let weight_dict: HashMap<H256, Weight> = weighted_validators
+            .iter()
+            .map(|v| (v.validator, v.weight))
+            .collect();
+
+        let mut cumulative_weight: u128 = 0;
+        let mut validators_included = HashSet::new();
```

Comment:
I actually don't think we need or want `validators_included` so that we can support a case where an ISM has duplicate validator addresses in there. Because we assume `weighted_validators` is accurate to what the ISM provides right

Reply:
I'm summing up in the weight_dict hashmap now, not sure that's what we want